mlx-rs/src/utils/guard.rs

```diff
@@ -0,0 +1,269 @@
+use half::{bf16, f16};
+use mlx_sys::{__BindgenComplex, bfloat16_t, float16_t, mlx_array};
+
+use crate::{complex64, error::{setup_mlx_error_handler, Exception}, Array};
+
+use super::SUCCESS;
+
+type Status = i32;
+
+pub trait Guard<T>: Default {
+    type MutRawPtr;
+
+    fn as_mut_raw_ptr(&mut self) -> Self::MutRawPtr;
+
+    fn set_init_success(&mut self, success: bool);
+
+    fn try_into_guarded(self) -> Result<T, Exception>;
+}
+
+pub(crate) trait Guarded: Sized {
```

Comment:
Took your approach and turned it into a trait, and this feels really nice right now